src/compiler/checker.ts

```diff
@@ -7555,6 +7559,62 @@ namespace ts {
             return undefined;
         }
 
+        function shouldAcquireLiteralType(literalNode: StringLiteral) {
+
+            let current: Node = literalNode;
+            while (true) {
+                const { parent } = current;
+                switch (parent.kind) {
+                    // The operand of a 'switch' should get a literal type.
+                    case SyntaxKind.SwitchStatement:
+                        return current === (parent as SwitchStatement).expression;
+
+                    // The tested expression of a 'case' clause should get a literal type.
+                    case SyntaxKind.CaseClause:
+                        return current === (parent as CaseClause).expression;
+
+                    case SyntaxKind.BinaryExpression:
+                        const binaryExpr = parent as BinaryExpression;
+                        switch (binaryExpr.operatorToken.kind) {
+                            // Either operand of an equality/inequality comparison
+                            // should get a literal type.
+                            case SyntaxKind.EqualsEqualsEqualsToken:
+                            case SyntaxKind.ExclamationEqualsEqualsToken:
+                            case SyntaxKind.EqualsEqualsToken:
+                            case SyntaxKind.ExclamationEqualsToken:
+                                return current === binaryExpr.left || current === binaryExpr.right;
+
+                            case SyntaxKind.BarBarToken:
+                                current = parent;
+                                continue;
+
+                        }
+                        break;

```

Comment:
I originally had it that way, but I thought it would be too indirect. I could go either way on the style, so I'll use change it back.